docs/sql/03-plan-caching.md

```diff
@@ -0,0 +1,132 @@
+# Plan Caching
+
+## Overview
+
+Optimization of a query may take considerable time. In many applications, the set of queries is fixed. Therefore, the result of
+query optimization could be cached and reused. This document explains the design of the plan cache in the Hazelcast Mustang SQL
+engine.
+
+In section 1 we discuss the high-level requirements to the plan cache. In section 2 we describe the design.
+
+## 1. Requirements
+
+We assume that the result of query optimization is deterministic. That is, the same query plan is produced for the same set of
+inputs. The inputs of the query optimizer are:
+
+- Catalog (schemas, tables, indexes)
+- The original query: query string, current schema, and parameters
+- Metadata
+
+### 1.1 Catalog
+
+A catalog is a set of objects that may participate in query execution. The catalog is often referred to as "schema" in
+the literature. We use the term "catalog" to disambiguate from the logical object containers, which are also called "schemas".
+
+The catalog has three types of objects. *Schema* is a logical container for other objects. *Table* is a relation backed
+by some physical storage, such as an `IMap`. *Index* is an additional data structure of a table that speeds up the execution
+of queries. The catalog is used to resolve objects mentioned in the query and choose the proper access method.
+
+If the catalog is changed, the plan created earlier might become invalid. For example, if the table is dropped, the execution
+of the plan will produce an error. If a new index is added, the optimizer may pick a better access path.
```

Comment:
some changes do not make the plan invalid but only less optimal (it is not guaranteed to be optimal anyway) - e.g. adding index. On the other hand, some changes cause that the plan cannot be used at all, eg. removing or altering referenced object